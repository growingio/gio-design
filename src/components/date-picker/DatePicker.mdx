import { Subtitle, ArgsTable, Story, Canvas } from '@storybook/addon-docs/blocks';
import DatePicker from '.';

# DatePicker 日期选择器

<Subtitle>选择一个日期</Subtitle>

## 参数说明

<ArgsTable of={DatePicker} />

## 代码演示

### 基础用法

<Canvas>
  <Story id="functional-components-datepicker--default" />
</Canvas>

### 选择范围

<!-- 这里无法添加 Canvas 包裹，否则会出现无限调用 DatePicker 组件函数进行渲染，导致堆栈溢出，原因未知 -->

<!-- <Canvas>
  <Story id="functional-components-datepicker--date-range-pickers" />
</Canvas>  -->

<Story id="functional-components-datepicker--date-range-pickers" />
