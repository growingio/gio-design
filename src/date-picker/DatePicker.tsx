/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from 'react';
import { usePrefixCls } from '@gio-design/utils';
import { LeftDoubleOutlined, LeftOutlined, RightOutlined, RightDoubleOutlined } from '@gio-design/icons';
import PickerPanel from 'rc-picker/lib/PickerPanel';
import generateDateFns from 'rc-picker/lib/generate/dateFns';
import zhCN from 'rc-picker/lib/locale/zh_CN';
import { DatePickerProps } from './interfaces';

function DatePicker(props: DatePickerProps) {
  return (
    <PickerPanel<Date>
      {...props}
      locale={zhCN}
      prefixCls={usePrefixCls('picker')}
      picker="date"
      generateConfig={generateDateFns}
      // @ts-ignore
      superPrevIcon={<LeftDoubleOutlined />}
      prevIcon={<LeftOutlined />}
      nextIcon={<RightOutlined />}
      superNextIcon={<RightDoubleOutlined />}
    />
  );
}

export default DatePicker;
