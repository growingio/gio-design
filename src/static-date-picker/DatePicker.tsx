import React, { useLayoutEffect, useRef } from 'react';
import { useControlledState, useLocale, usePrefixCls } from '@gio-design/utils';
import { LeftDoubleOutlined, LeftOutlined, RightOutlined, RightDoubleOutlined } from '@gio-design/icons';
import PickerPanel from 'rc-picker/lib/PickerPanel';
import generateDateFns from 'rc-picker/lib/generate/dateFns';
import { Locale } from 'rc-picker/lib/interface';
import { isEqual } from 'date-fns';
import defaultLocale from './locales/zh-CN';
import { StaticDatePickerProps } from './interfaces';

const OmittedCell: React.FC = () => {
  const spanRef = useRef<HTMLSpanElement>(null);
  useLayoutEffect(() => {
    // Remove title prop
    if (spanRef.current) {
      const parent = spanRef.current.parentElement;
      if (parent?.hasAttribute('title')) {
        parent.removeAttribute('title');
      }
    }
  }, []);
  return <span ref={spanRef} />;
};

const DatePicker: React.FC<StaticDatePickerProps> = ({
  viewDate: viewDateProp,
  disabledDate: disabledDateProp,
  ...restProps
}) => {
  const locale = useLocale<Locale>('DatePicker') || defaultLocale;
  const [viewDate, setViewDate] = useControlledState(viewDateProp, new Date());

  const prefixCls = usePrefixCls('picker');

  const isSameYearAndDay = (currentDate: Date) =>
    isEqual(
      new Date(currentDate.getFullYear(), currentDate.getMonth()),
      new Date(viewDate.getFullYear(), viewDate.getMonth())
    );

  const omitOtherDate = (currentDate: Date) => {
    if (isSameYearAndDay(currentDate)) {
      return <div className={`${prefixCls}-cell-inner`}>{currentDate.getDate()}</div>;
    }
    return <OmittedCell />;
  };

  const disabledDate = (currentDate: Date) => {
    if (!isSameYearAndDay(currentDate)) {
      return true;
    }
    if (typeof disabledDateProp === 'function') {
      return disabledDateProp(currentDate);
    }
    return false;
  };

  return (
    <PickerPanel<Date>
      data-testid="static-date-picker"
      dateRender={omitOtherDate}
      disabledDate={disabledDate}
      {...restProps}
      pickerValue={viewDate}
      onPickerValueChange={(date) => setViewDate(date)}
      locale={locale}
      prefixCls={prefixCls}
      picker="date"
      generateConfig={generateDateFns}
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      superPrevIcon={<LeftDoubleOutlined />}
      prevIcon={<LeftOutlined />}
      nextIcon={<RightOutlined />}
      superNextIcon={<RightDoubleOutlined />}
    />
  );
};

DatePicker.displayName = 'DatePicker';

export default DatePicker;
