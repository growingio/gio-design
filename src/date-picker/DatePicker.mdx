import { Subtitle, ArgsTable, Story, Canvas } from '@storybook/addon-docs';
import DatePicker from './index';

# DatePicker 日期选择器

<Subtitle>当用户需要一个日期，可以在面板中进行选择。</Subtitle>

## 代码演示

### 基本样式

最简单的用法，在面板中可以选择日期。

<Canvas>
  <Story id="pickers-datepicker--basic" />
</Canvas>

### 禁止选择部分日期

可用 `disabledDate` 禁止选择部分日期.

<Canvas>
  <Story id="pickers-datepicker--disabled-date" />
</Canvas>

## 参数说明

<ArgsTable of={DatePicker} />
