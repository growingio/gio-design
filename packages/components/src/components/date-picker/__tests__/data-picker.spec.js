import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount, render } from 'enzyme';
import moment from 'moment';
import { noop } from 'lodash';
import DatePicker from '..';

const format = 'YYYY/MM/DD';
const VALUE = moment([2015, 5, 1]);
const disabledDate = (value) => {
  const date = moment();
  date.hour(0);
  date.minute(0);
  date.second(0);
  return value.isBefore(date); // can not select days before today
};

async function waitComponentRender(component, amount = 3000) {
  await act(async () => new Promise((resolve) => setTimeout(resolve, amount)).then(() => component.update()));
}

describe('DatePicker', () => {
  const DatePickerInstance = () => (
    <DatePicker
      value={VALUE}
      onChange={noop}
      onSelect={noop}
      format={format}
      showFooter
      disabledDate={disabledDate}
    />
  );

  it('should match snapshot', () => {
    const wrapper = render(DatePickerInstance());
    expect(wrapper).toMatchSnapshot();
  });

  it('should receive props', () => {
    expect(() => {
      const wrapper = mount(DatePickerInstance());
      wrapper.setProps({ disabledDate: noop });
      wrapper.setProps({ format });
      wrapper.setProps({ showFooter: 'true' });
      wrapper.unmount();
    }).not.toThrow();
  });

  it('should popup corretly', (done) => {
    const wrapper = mount(DatePickerInstance());
    wrapper.find('.gio-input-content').simulate('click');
    waitComponentRender(wrapper).then(() => {
      expect(wrapper.exists('.gio-input-content')).toBe(true);
      done();
    });
  });

  it('should render corretly', () => {
    const wrapper = mount(DatePickerInstance());
    const content = wrapper.find('.gio-input-content');
    expect(content.html().includes('2015/05/01'));
  });
});
