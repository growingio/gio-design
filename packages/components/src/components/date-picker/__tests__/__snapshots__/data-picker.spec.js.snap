// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker should match snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "gio-date-picker-wrap",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "gio-input",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-input-content gio-input-content-middle",
              "placeholder": "请输入…",
              "type": "text",
              "value": "2015/06/01",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-date-picker-wrapper",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-date-picker-wrapper",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-input",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-input-content gio-input-content-middle",
                "placeholder": "请输入…",
                "type": "text",
                "value": "2015/06/01",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
